File: frontend/src/utils/price.ts
/**
 * Price Parsing Utilities
 *
 * Extracts numeric values from string-based price fields (e.g. "$249.99", "€120", "1,250.00").
 * Mirrors the backend parse_price() utility in backend/app/utils/price.py.
 */

/**
 * Extract a numeric value from a price string.
 * Handles formats: "$249.99", "€120", "150 USD", "1,250.00", "1.250,00"
 */
export function parsePrice(priceStr: string | null | undefined): number | null {
  if (!priceStr) return null

  // Strip currency symbols and letters
  let cleaned = priceStr.replace(/[^0-9.,\-]/g, '').trim()
  if (!cleaned) return null

  // Detect European format: "1.250,00" (dot as thousands, comma as decimal)
  if (/^\d{1,3}(\.\d{3})+(,\d{1,2})?$/.test(cleaned)) {
    cleaned = cleaned.replace(/\./g, '').replace(',', '.')
  }
  // Handle comma as decimal: "250,00"
  else if (/^\d+(,\d{1,2})$/.test(cleaned)) {
    cleaned = cleaned.replace(',', '.')
  }
  // Handle comma as thousands: "1,250.00"
  else {
    cleaned = cleaned.replace(/,/g, '')
  }

  const value = parseFloat(cleaned)
  return isNaN(value) ? null : value
}

/**
 * Format a number as a price string.
 */
export function formatPrice(amount: number, currency = 'EUR'): string {
  try {
    return new Intl.NumberFormat(undefined, {
      style: 'currency',
      currency,
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    }).format(amount)
  } catch {
    return `${currency} ${amount.toFixed(2)}`
  }
}

File: backend/app/utils/price.py
"""
Price Parsing Utility

Extracts numeric values from price strings that may contain
currency symbols, codes, or formatting.
"""
import re
from typing import Optional


def parse_price(price_str: Optional[str]) -> Optional[float]:
    """Extract numeric value from a price string.

    Examples:
        parse_price("$249.99") -> 249.99
        parse_price("€120") -> 120.0
        parse_price("150 USD") -> 150.0
        parse_price("1,250.00") -> 1250.0
        parse_price(None) -> None
        parse_price("N/A") -> None
    """
    if not price_str:
        return None

    # Remove everything except digits, dots, commas, and minus
    cleaned = re.sub(r'[^\d.,\-]', '', price_str)
    if not cleaned:
        return None

    # Handle mixed dot/comma formats
    if ',' in cleaned and '.' in cleaned:
        if cleaned.rfind(',') > cleaned.rfind('.'):
            # European: 1.250,00
            cleaned = cleaned.replace('.', '').replace(',', '.')
        else:
            # American: 1,250.00
            cleaned = cleaned.replace(',', '')
    elif ',' in cleaned:
        parts = cleaned.split(',')
        if len(parts) == 2 and len(parts[1]) <= 2:
            # Likely decimal: 100,50
            cleaned = cleaned.replace(',', '.')
        else:
            # Likely thousands: 1,250
            cleaned = cleaned.replace(',', '')

    try:
        return float(cleaned)
    except ValueError:
        return None
